File: Python/Data Structure and Algorithms/Searching/Binary Search/InterviewBit/Search In Bitonic Array.py

def bin_search(A, B, l, r):
    if A[l] < A[r]:
        
        while l <= r:
            mid = int((l+r)/2)
            if B == A[mid]:
                return mid
            elif B > A[mid]:
                l = mid+1
            else: 
                r = mid - 1
    else:
        
        while l <= r:
            mid = int((l+r)/2)
            if B == A[mid]:
                return mid
            elif B > A[mid]:
                r = mid-1
            else: 
                l = mid + 1
    return -1

class Solution:
    # @param A : list of integers
    # @param B : integer
    # @return an integer
    

    def solve(self, A, B):
        l = 0
        r = len(A) - 1
        while l <= r:
            mid = int((l+r)/2)
            if (mid+1) < len(A) and (mid-1) >=0:
                if A[mid] > A[mid-1] and A[mid] > A[mid+1]:
                    ans1 = bin_search(A, B, 0, mid)
                    ans2 = bin_search(A, B, mid+1, len(A)-1)
                    if ans1 == -1 and ans2 == -1:
                        return -1
                    elif ans1 == -1:
                        return ans2
                    else:
                        return ans1
                elif A[mid] > A[mid-1] and A[mid] < A[mid+1]:
                    l = mid+1
                elif A[mid] < A[mid-1] and A[mid] > A[mid+1]:
                    r = mid-1

File: C++/Data Structure and Algorithms/Searching/Binary Search/InterviewBit/Search In Bitonic Array.cpp
int bin_search(vector<int> &A, int B, int l, int r)
{
    if(A[l] < A[r])
    {
        while(l <= r)
        {
            int mid = (l+r)/2;
            if(B == A[mid])
            {
                return mid;
            }
            else if(B > A[mid])  l = mid+1;
            else r = mid -1;
        }
    }
    else
    {
        while(l <= r)
        {
            int mid = (l+r)/2;
            if(B == A[mid])
            {
                return mid;
            }
            else if(B > A[mid])  r = mid-1;
            else l = mid+1;
        }
    }
    return -1;
}

int Solution::solve(vector<int> &A, int B)
{
    int l = 0;
    int r = A.size() - 1;
    while(l <= r)
    {
        int mid = (l+r)/2;
        if(A[mid] > A[mid-1] && A[mid] > A[mid+1])
        {
            int ans1 = bin_search(A, B, 0, mid);
            int ans2 = bin_search(A, B, mid+1, A.size()-1);
            if(ans1 == -1 && ans2 == -1) return -1;
            else if(ans1 == -1) return ans2;
            else return ans1;
        }
        else if(A[mid] > A[mid-1] && A[mid] < A[mid+1]) l = mid+1;
        else if(A[mid] < A[mid-1] && A[mid] > A[mid+1]) r = mid-1;
    }
}